### PR TITLE
Enforce the rules applied to the handler method annotations

### PR DESCRIPTION
**Pull request by armiol**
This changeset updates the rules applied to the signatures of handler methods. 

In particular, methods annotated with `@Assign`, `@React`, `@Command` and `@Subscribe` are carefully checked for the requirements set for each of the annotations. This behavior has been documented and covered with unit tests.

Special `DoNothing` command message has been introduced. It should be used as a return result of `@Command`-ing methods if they declare an option not to emit any command in response to the incoming event:

```java
@Command
EitherOf3<PauseTask, StopTask, DoNothing> on(TaskUnassigned event) {...}
```

Such a command message is a counterpart of `Nothing` event message. Similarly, it is ignored by the framework once emitted.

The framework version is set to `1.1.6`.

### CODE REVIEW

File: server/src/main/java/io/spine/server/command/model/CommandAcceptingSignature.java [hunk 1]
```diff
@@ -79,7 +88,7 @@
         MESSAGE {
             @Override
             public boolean matches(MethodParams params) {
-                return params.is(CommandMessage.class);
+                return params.is(CommandMessage.class) && params.is(GeneratedMessageV3.class);
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/command/model/CommandAcceptingSignature.java]: Why `GeneratedMessageV3`? I suppose, `Message` provides a sufficient interface. And `CommandMessage` extends `Message`.

File: server/src/main/java/io/spine/server/entity/EntityVisibility.java [hunk 2]
```diff
@@ -84,22 +87,43 @@ private EntityVisibility(Visibility value) {
...
+     *         into this method.
+     *         We don't want to define {@code (entity)} option for the {@code Event} message,
+     *         therefore this method handles {@code Event.class} as a special case and returns
+     *         {@code NONE} visibility level for it.
      */
-    public static EntityVisibility of(Class<? extends Message> stateClass) {
+    public static Optional<EntityVisibility> of(Class<? extends Message> stateClass) {
         checkNotNull(stateClass);
+
+        if(Event.class.equals(stateClass)) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/EntityVisibility.java]: Please fix formatting.

File: server/src/main/java/io/spine/server/event/model/StateSubscriberSpec.java [hunk 3]
```diff
@@ -68,7 +70,11 @@ public boolean matches(MethodParams params) {
         }
         @SuppressWarnings("unchecked") // Checked above.
         Class<? extends Message> firstParameter = (Class<? extends Message>) params.type(0);
-        EntityVisibility visibility = EntityVisibility.of(firstParameter);
+        Optional<EntityVisibility> visibilityOption = EntityVisibility.of(firstParameter);
+        if(!visibilityOption.isPresent()) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/event/model/StateSubscriberSpec.java]: Please fix formatting.

File: server/src/main/java/io/spine/server/model/MethodParams.java [hunk 4]
```diff
@@ -113,6 +113,16 @@ public boolean is(Class<?> type) {
...
             return false;
         }
+        return firstIs(type);
+    }
+
+    /**
+     * Verifies if the first parameter of the method is of the passed type.
+     */
+    public boolean firstIs(Class<?> type) {
+        if(size() == 0) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/model/MethodParams.java]: Same here.

File: server/src/test/java/io/spine/server/model/MethodSignatureTest.java [hunk 5]
```diff
@@ -55,11 +60,57 @@ protected final void testValid(Method method) {
     @ParameterizedTest
     @MethodSource("invalidMethods")
     protected final void testInvalid(Method method) {
-        assertThrows(SignatureMismatchException.class, () -> wrap(method));
+
+        //TODO:2019-09-24:alex.tymchenko: a workaround to detect invalid `@Command`-er methods.
```
Comment (dmdashenkov) [inline server/src/test/java/io/spine/server/model/MethodSignatureTest.java]: Please resolve.

File: server/src/test/java/io/spine/server/model/MethodSignatureTest.java [hunk 6]
```diff
@@ -55,11 +60,57 @@ protected final void testValid(Method method) {
...
+        try {
+            Optional<? extends HandlerMethod> result = wrap(method);
+            if(result.isPresent()) {
```
Comment (dmdashenkov) [inline server/src/test/java/io/spine/server/model/MethodSignatureTest.java]: Please fix formatting.

**Comment by armiol** [comment]
@dmdashenkov comments addressed. PTAL again.

File: server/src/main/java/io/spine/server/command/Command.java [hunk 7]
```diff
@@ -171,7 +171,9 @@
  * <pre>
  *
  * {@literal @}Command
- *  CreateProject on(Rejections.CannotCreateProject rejection) { {@literal // change the parameters and try again...} }
+ *  CreateProject on(Rejections.CannotCreateProject rejection) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/command/Command.java]: Please check identation here as well.

File: server/src/main/java/io/spine/server/command/model/CommandReactionSignature.java [hunk 8]
```diff
@@ -122,7 +121,7 @@ public boolean matches(MethodParams params) {
             @Override
             public boolean matches(MethodParams params) {
                 return params.are(EventMessage.class, EventContext.class)
-                        && params.firstIs(GeneratedMessageV3.class);
+                        && params.declaredAsClasses();
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/command/model/CommandReactionSignature.java]: Can we check that all params are declared as classes in a single place? Now, we have to copy this check for each signature.

**Pull request merged**